vtrace/envitools.py

```diff
@@ -25,36 +36,134 @@ def cmpRegs(emu, trace):
     return True
 
 
-def emuFromTrace(trace):
-    '''
-    Produce an envi emulator for this tracer object.
-    '''
-    arch = trace.getMeta('Architecture')
-    plat = trace.getMeta('Platform')
-    amod = envi.getArchModule(arch)
-    emu = amod.getEmulator()
+skip_mem_opcodes_by_arch = {
+        'i386': (e_i386const.INS_NOP, e_i386const.INS_LEA),
+        'amd64': (e_i386const.INS_NOP, e_i386const.INS_LEA),
+        }
+
+
+class LockstepEmulator:
+    def __init__(self, trace):
+        self.trace = trace
+        self.emu = vutil.emuFromTrace(trace)    # this should set up the emulator
+        self.arch = self.trace.getMeta("Architecture")
+
+    def go(self):
+        count = 0
+        skip_mem_opcodes = skip_mem_opcodes_by_arch.get(self.arch, ())
+
+        self.cmpRegs()
+        while True:
+            print("%4d  Lockstep: 0x%.8x" % (count, self.emu.getProgramCounter()), end='')
+            try:
+                pc = self.trace.getProgramCounter()
+                op = self.emu.parseOpcode(pc)
+                print("  %r  " % op, end='')
+
+                # don't compare memory from these instruction results (error prone)
+                skip_mem = False
+                if op.opcode in skip_mem_opcodes:
+                    skip_mem = True
+
+                # store the current opcode addresses for comparison later (for when the 
+                #  original register gets clobbered
+                if not skip_mem:
+                    self.grabMemAddrs(op)
+
+                for oper in op.opers:
+                    if skip_mem:
+                        continue
+                    print("(%r = %x)" % (oper.repr(op), oper.getOperValue(op, self.emu)), end='')
+
+                self.trace.stepi()
+                self.emu.stepi()
+                count += 1
+
+                if op.mnem == 'rdtsc':
+                    self.syncRegsFromTrace()
+
+                print("(eflags: 0x%x)" % self.trace.getRegisterByName('eflags'), end='')
+                print("(eflags: 0x%x)" % self.emu.getRegisterByName('eflags'), end='')
+                self.cmpRegs()
+
+                if not skip_mem:
+                    self.cmpMem(op)
+
+            except RegisterException as msg:
```

Comment:
Hmm. I feel like we should flesh this portion out. Have a better REPL style workflow, not just for errors, but in general so we can step through each instruction and examine things as they go.

Just a thought. Kind of a big request that we can talk through, but I think it'd be useful.

Reply:
i'd love the conversation.  i like mine for some things (interactive), but i could be convinced that yours is better.

Reply:
i love the REPL approach.  let's make this a different PR.